salt/pillar/nsot.py

```diff
@@ -0,0 +1,181 @@
+# -*- coding: utf-8 -*-
+'''
+A module that adds data to the Pillar structure from the NSoT API.
+
+.. versionadded:: Neon
+
+Configuring the NSoT ext_pillar
+---------------------------------
+
+The following fields are required:
+
+.. code-block:: yaml
+
+    ext_pillar:
+      - nsot:
+        api_url: http://nsot_url.com/api/
+        email: 'user@site.com'
+        secret_key: abc123
+
+.. note::
+
+    Note the trailing slash on the api_url field. Although NSoT gives the option
+    of whether or not to use the tailing slash, this module assumes that the
+    APPEND_SLASH option is set to True (the default)
+
+The following fields are optional:
+
+replace_dot: ``str``
+    This is in case a minion_id contains a dot. Because NSoT
+    doesn't allow hostnmes with dots, this option allows you to look up a device
+    with this character replacing dots. For example, with this option set to
+     '-', the minion rtr1.lax1 would be looked up in NSoT as rtr1-lax1.
+
+all_device_info: ``list``
+    This is a list of minions that will receive all device
+    information in NSoT. For example, if this list contained the entry
+    'test_min', then the minion 'test_min' would have all devices in NSoT under
+    pillar['nsot']['devices']
+
+minion_regex: ``list``
+    This is in case you only want certain minions being looked
+    up in NSoT. For example, if you have 500 minions on a single master, you may
+    not want all of them hitting the NSoT endpoint. With this list of regexes,
+    only minions matching the regex will be queried.
+
+Here's an example config with all options:
+
+.. code-block:: yaml
+
+    ext_pillar:
+      - nsot:
+        api_url: http://nsot_url.com/api/
+        email: 'user@site.com'
+        secret_key: abc234
+        replace_dot: '-'
+        all_device_info:
+          - minion_1
+          - minion_2
+        minion_regex:
+          - 'rtr*'
+          - 'sw*'
+          - '^router*'
+
+'''
+
+from __future__ import absolute_import, print_function, unicode_literals
+import logging
+import re
+
+# Import Salt libs
+import salt.utils.http
+
+log = logging.getLogger(__name__)
+
+
+def _get_token(url, email, secret_key):
+    '''
+    retrieve the auth_token from nsot
+
+    :param url: str
+    :param email: str
+    :param secret_key: str
+    :return: str
+    '''
+    url += 'authenticate/'
+    data_dict = {"email": email, "secret_key": secret_key}
+    query = salt.utils.http.query(url, data=data_dict, method='POST',
+                                  decode=True)
+    error = query.get('error')
+    if error:
+        log.error('can\'t get auth_token from nsot! reason: %s', error)
```

Comment:
What is "auth_token"? This is just a name of a key inside the structure inside JSON that end user should never even care about. Please do not put such stuff into the logs. Instead please rephrase it as it would be non-Salt-aware guy reading it, and also write it properly, from capital letter, something like "Unable to obtain authentication token from NSoT: %s".

Reply:
An auth_token is generated based on a valid email/secret_key. If it's unable to be obtained, the end user should care about it since we're expecting to obtain an auth_token if the proper creds are provided and the service is running.

Reply:
This was a clue question. :wink: My point is that an ordinary newcomer will just stare at this message, having no idea where this coming from, why and what to do next. So you'd better to be a bit more verbose here and rephrase it with generic terminology so it is clear what action to take next.